File: runtimes/go/crypto/kms.go
package crypto

import (
	"context"

	"github.com/aws/aws-sdk-go-v2/service/kms"
)

// HelixKMS implements AWS KMS payload encryption and decryption.
type HelixKMS struct {
	client *kms.Client
	keyID  string
}

// NewHelixKMS creates a new HelixKMS instance.
func NewHelixKMS(client *kms.Client, keyID string) *HelixKMS {
	return &HelixKMS{
		client: client,
		keyID:  keyID,
	}
}

// EncryptPayload encrypts a payload using AWS KMS.
func (h *HelixKMS) EncryptPayload(ctx context.Context, payload []byte) ([]byte, error) {
	input := &kms.EncryptInput{
		KeyId:     &h.keyID,
		Plaintext: payload,
	}
	output, err := h.client.Encrypt(ctx, input)
	if err != nil {
		return nil, err
	}
	return output.CiphertextBlob, nil
}

// DecryptPayload decrypts a payload using AWS KMS.
func (h *HelixKMS) DecryptPayload(ctx context.Context, ciphertext []byte) ([]byte, error) {
	input := &kms.DecryptInput{
		KeyId:          &h.keyID,
		CiphertextBlob: ciphertext,
	}
	output, err := h.client.Decrypt(ctx, input)
	if err != nil {
		return nil, err
	}
	return output.Plaintext, nil
}

File: runtimes/rust/src/crypto/kms.rs
use aws_sdk_kms::Client;
use aws_sdk_kms::primitives::Blob;
use std::sync::Arc;

#[derive(Clone)]
pub struct HelixKMS {
    client: Arc<Client>,
    key_id: String,
}

impl HelixKMS {
    pub async fn new(key_id: String) -> Self {
        let config = aws_config::load_from_env().await;
        let client = Arc::new(Client::new(&config));
        Self { client, key_id }
    }

    pub async fn encrypt_payload(&self, plaintext: &[u8]) -> Result<Vec<u8>, aws_sdk_kms::Error> {
        let resp = self
            .client
            .encrypt()
            .key_id(&self.key_id)
            .plaintext(Blob::new(plaintext.to_vec()))
            .send()
            .await?;
        
        Ok(resp.ciphertext_blob().unwrap().clone().into_inner())
    }

    pub async fn decrypt_payload(&self, ciphertext: &[u8]) -> Result<Vec<u8>, aws_sdk_kms::Error> {
        let resp = self
            .client
            .decrypt()
            .key_id(&self.key_id)
            .ciphertext_blob(Blob::new(ciphertext.to_vec()))
            .send()
            .await?;
            
        Ok(resp.plaintext().unwrap().clone().into_inner())
    }
}
